### PR TITLE
Add authentication options to Webhook

### PR DESCRIPTION
**Pull request by fhscholl**
#### Database Migration
NO

#### Description
Add the fields and functionality to Webhook to allow APIs that require authentication.

#### Todos
- [ ] Tests
- [ ] Documentation

#### Issues Fixed or Closed by this PR

* #2257

### CODE REVIEW

**Comment by fhscholl** [comment]
Not sure what's up with the test failure. The test passes when I run it locally.

File: src/NzbDrone.Core/Notifications/Webhook/WebhookProxy.cs [hunk 1]
```diff
@@ -30,6 +32,11 @@ public void SendWebhook(WebhookPayload body, WebhookSettings settings)
                 request.Headers.ContentType = "application/json";
                 request.SetContent(body.ToJson());
 
+                if (!String.IsNullOrEmpty(settings.Username) || !String.IsNullOrEmpty(settings.Password))
+                {
+                    var networkCredential = new NetworkCredential(settings.Username, settings.Password);
```
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Webhook/WebhookProxy.cs]: This is unnecessary as the username and password are pulled out of the network credential in `SetNetworkCredential`

File: src/NzbDrone.Common/Http/HttpRequest.cs [hunk 2]
```diff
@@ -76,5 +77,15 @@ public void SetContent(string data)
             var encoding = HttpHeader.GetEncodingFromContentType(Headers.ContentType);
             ContentData = encoding.GetBytes(data);
         }
+
+        public void SetNetworkCredential(NetworkCredential networkCredential)
```
Comment (markus101) [inline src/NzbDrone.Common/Http/HttpRequest.cs]: How about `public void AddBasicAuthentication(string username, string password)`?

File: src/NzbDrone.Common/Http/HttpRequest.cs [hunk 3]
```diff
@@ -76,5 +77,15 @@ public void SetContent(string data)
...
+        {
+            if (networkCredential != null)
```
Comment (markus101) [inline src/NzbDrone.Common/Http/HttpRequest.cs]: This should never be called if network credential is null, if it is, it should explode.

File: src/NzbDrone.Core/Notifications/Webhook/WebhookProxy.cs [hunk 4]
```diff
@@ -30,6 +31,10 @@ public void SendWebhook(WebhookPayload body, WebhookSettings settings)
                 request.Headers.ContentType = "application/json";
                 request.SetContent(body.ToJson());
 
+                if (!String.IsNullOrEmpty(settings.Username) || !String.IsNullOrEmpty(settings.Password))
```
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Webhook/WebhookProxy.cs]: `if (settings.Username.IsNotNullOrWhiteSpace() || settings.Password.IsNotNullOrWhiteSpace())`

File: src/NzbDrone.Common/Http/HttpRequest.cs [hunk 5]
```diff
@@ -76,5 +76,12 @@ public void SetContent(string data)
             var encoding = HttpHeader.GetEncodingFromContentType(Headers.ContentType);
             ContentData = encoding.GetBytes(data);
         }
+
+        public void AddBasicAuthentication(string username, string password)
+        {
+            var authInfo = username + ":" + password;
+            authInfo = Convert.ToBase64String(Encoding.GetEncoding("ISO-8859-1").GetBytes(authInfo));
```
Comment (markus101) [inline src/NzbDrone.Common/Http/HttpRequest.cs]: `var authInfo = Convert.ToBase64String(Encoding.GetEncoding("ISO-8859-1").GetBytes($"{username}:{password}"));`

**Comment by markus101** [comment]
Merged after some changes.

**Pull request closed without merging**